iterated-prisoners-dilemma/aniket/evolution.py

```diff
@@ -0,0 +1,129 @@
+       chosen by the two players.
+       bit value of 1 --> cooperation
+       bit value of 2 --> defection"""
+
+    if (bit_a,bit_b)==('1','1'):
+        return (3,3)
+    elif (bit_a,bit_b)==('1','0'):
+        return (0,5)
+    elif (bit_a,bit_b)==('0','1'):
+        return (5,0)
+    else:
+        return (1,1)  
+
+def get_strategy(encoding):
+    """return the index in the strategy 
+       corresponding to the initial memory of the
+       3 games"""
+
+    return int(encoding, 2)
+
+def play_game(strategy_a, strategy_b):
+    """we have a total of 100 games between two strategies"""
+    
+    score_a = score_b = 0
+    encoding_a = strategy_a[64:70]
+    encoding_b = strategy_b[64:70]
+    for i in range (0, 100):
+        move_a = strategy_a[get_strategy(encoding_a)]
+        move_b = strategy_b[get_strategy(encoding_b)]
+        a, b = assign_score(move_a, move_b)
+        score_a += a
+        score_b += b
+        encoding_a = encoding_a[2:]+ move_a + move_b
+        encoding_b = encoding_b[2:]+ move_b + move_a
+    #averaging the scores over 100 games        
+    score_a/=100
+    score_b/=100        
+    return (score_a, score_b)    
+                
+
+def fitness_calculation(strategies, fitness):
+    """each strategy plays with all other 19 strategies 
+       and with itself i.e. a total of 20 games"""
+
+    for i in range(0, 20):
+        for j in range(i, 20):
+            x, y = play_game(strategies[i], strategies[j])
+            fitness[i]+=x
+            fitness[j]+=y
+    #averaging the fitness values over 20 strategies            
+    fitness = [f/20 for f in fitness]
+    return fitness
+
+def fitness_proportionate_selection(strategies, fitness):
+    """choosing parent strategies on the basis of roulette 
+       wheel selection, and we perform this process for 'n'
+       number of times, where 
+       n --> number of strategies, here 20."""
+
+    max_val = sum(fitness)
+    parents = []
+    size = len(fitness)
+    for i in range(size):
+        temp_sum = 0
+        rand_val = random.uniform(0, max_val)
+        for j in range(size):
+            temp_sum += fitness[j]
+            if temp_sum>=rand_val:
+                parents+=[strategies[j]]
+                break
+    return parents
+
+def mutate(offspring_1, pm):
+    """simple mutation of the bits of the offspring
+       with repsective probability"""
+
+    for i in range(len(offspring_1)):
+        if random.random()<=pm:
+            offspring_1 = offspring_1[:i]+str(1^int(offspring_1[i]))+offspring_1[i+1:]
+    return offspring_1           
+
+def crossover(parent_1, parent_2):
+    """a simple crossover between the two parents
+       by choosing a random crossover point."""
+
+    size = len(parent_1)
+
+    #choosing a crossover point
+    crossover_point = random.randint(0, size-1)
+    if crossover_point==0:
+        return [parent_2, parent_1]
+    elif crossover_point==size-1:
+        return [parent_1, parent_2]
```

Comment:
In the current implementation, crossover doesn't occur with probability pc. It occurs with probability pc * (size-2)/size (because of the above two if statements)

Also, if the crossover_point is 0, then a crossover will occur. For e.g.,
take the chromosomes 0100 and 1000 and crossover_point = 0 then,
the offsprings will be 0000 and 1100.

Reply:
I have made the requested changes :smiley: